### ISSUE TITLE
Inform user when loading ontology fails

### ISSUE THREAD

**Issue by secrashi** [edamontology/edam-browser#137]
I tried to load a custom ontology it was not functional as shown in the video attached below: 

https://user-images.githubusercontent.com/56053777/115600795-af35e280-a2fa-11eb-8e67-5842bd668d26.mp4

Can @matuskalas @hmenager @bryan-brancotte assign me this so I'll start working on it asap!

**Comment by tawahpeggy** [comment]
Hello @secrashi, I noticed the same problem and think you should move on with your solution for it. And open a pull request, they will check when they are less busy. sorry about any inconvenience.

**Comment by matuskalas** [comment]
Thank you very much @secrashi and @tawahpeggy for the heads up 🙏🏽

This really is a UX problem, where it's not clear enough from the instructions to a novice user, that the URL or file uploaded have to be in a special JSON format in order to work. It's written there in detail, but only underneath the active boxes. Working on a better input format (JSON-LD directly, as mentioned elsewhere), will definitely help here (but I'm not sure how big task that is).

For now, if you'd like to slightly improve the instructions there to make it more clear, then great, please go ahead. But if you're working on more interesting things then please feel free to leave this 😊

I'm changing the title of this issue accordingly...

**Comment by secrashi** [comment]
Ohh, Thanks @matuskalas for letting me know, I would soon make a PR regarding the instructions. Also, this made me wonder what else we can do with this:
Can we add in a new feature, in which we can select a term and download it in JSON-LD format. would it be useful?

**Comment by matuskalas** [comment]
> Ohh, Thanks @matuskalas for letting me know, I would soon make a PR regarding the instructions. 

Thank you so much! 👍🏽 I wondered what would be the best & easiest: just putting the details text BEFORE the input fields, or adding a note to the inputs that it needs to be in JSON, or something better? But you might have a better idea, so I'll leave the decision to you, you know better how it feels to a relatively new user😉

> Can we add in a new feature, in which we can select a term and download it in JSON-LD format. would it be useful?

It would definitely be very useful! 🙌🏽 But:
- It might be a bigger task to implement
- We (all together) need to consider whether it fits as a feature to EDAM Browser or should rather be implemented directly at _edamontology.org_

Maybe a natural succession could be first implementing JSON-LD as an input, and then designing a solution for JSON-LD "snippets" for single concepts and an API to provide them. In any case, if you have any good ideas, we're eager to hear|see them! 😊

**Comment by secrashi** [comment]
> Thank you so much! 👍🏽 I wondered what would be the best & easiest: just putting the details text BEFORE the input fields, or adding a note to the inputs that it needs to be in JSON, or something better? But you might have a better idea, so I'll leave the decision to you, you know better how it feels to a relatively new user😉

- I'm planning to put the instructions in the placeholder of input. That way it would be clearly visible to everyone. 
- Secondly, we can perform a check on the input file/ link, whether it contains data in JSON-LD format or not. If not, then it'll throw an error and ask the user to upload it in JSON-LD format. I looked up ways to do this, and I found that there are two methods for this, 

1.  To use IsValid(JToken, JsonSchema) method with the JSON Schema.
2.  Use JSON parsing library 

I will try both of these methods, and see if any of these works. 



> * It might be a bigger task to implement (unless you have an idea about a relatively simple solution)
> * We (all together) should consider whether it fits as a feature to EDAM Browser or should rather be implemented directly at _edamontology.org_ (although perhaps it doesn't matter so much)
> 
> Maybe a natural succession could be first implementing JSON-LD as an input and then designing a solution for JSON-LD "snippets" for single concepts and an API to provide them. (Again, unless you have some elegant solution in mind.) In any case, if you have any good ideas, we're eager to hear|see them! 😊

- We have various functions and npm package to convert text and markdown files into JSON format. Also, we can use Rasa since it is comparatively easier to use it convert data in JSON format { I found it over the internet}. We can use this both for single concepts as well as file uploads.
- Our main challenge if we implement this feature would be to identify the file type user uploaded and then proceed to convert it into JSON. For a start, we can specify a limited number of file types that the user can upload, and then continue adding more file types under the belt. 

I'm very curious about how these methods work and what all ways we can alter them to do what's required. Though I agree, this would be a bigger task to implement but if @matuskalas approve of this, I would love to work over this.

**Comment by bryan-brancotte** [comment]
Hi @secrashi 

I only saw your gitter message at first but not the issue before I pushed 1af14999e6934c89b751af4716128ec80b2167a4 which bring some help in the current state of the application. Feel free to improve.

The main issue here is that you provided an url to a term in a field of a form that was waiting for a whole ontology in a precise format. If you look at the console (a good habit as a dev when a button doesn't work as expected) you can see that something went wrong as an error stack trace is displayed:
```
Uncaught TypeError: Cannot read property 'meta' of null
    at tree-reusable-d3.js:808
    at Object.<anonymous> (d3.min.js:2)
    at k.call (d3.min.js:2)
    at XMLHttpRequest.e (d3.min.js:2)
```
The code says that object `json` in [tree-reusable-d3.js:808](https://github.com/edamontology/edam-browser/blob/main/js/tree-reusable-d3.js#L808) is null which means that nothing was parse from the given URL. As an UX improvement we can notify the user that nothing was loaded.

Having the possibility to download the term in json/json-ld is a whole new topic and should be treated separately.

**Comment by bryan-brancotte** [comment]
_this issue is a duplicate of #137, keeping this open as a discussion took place here_

**Comment by secrashi** [comment]
> The main issue here is that you provided an url to a term in a field of a form that was waiting for a whole ontology in a precise format. If you look at the console (a good habit as a dev when a button doesn't work as expected) you can see that something went wrong as an error stack trace is displayed:

Yes, @bryan-brancotte, I looked at those error messages, but as a user, if he/she enters a wrong link, then nothing should be loaded and an error message should appear. But instead, on loading custom ontology while giving the wrong link( one without json-ld format ), it shows up some changes in the website as shown in the figure below:

![WhatsApp Image 2021-04-27 at 8 53 54 AM](https://user-images.githubusercontent.com/56053777/116180686-f971f600-a736-11eb-9d2b-7c802857b31c.jpeg)

The colour of branches changes and the right section( details + previously seen terms) disappears! So, we need to change the code such that it first verifies the link and then make any changes! That's why I insisted on performing a check on the input file/ link.


> Having the possibility to download the term in json/json-ld is a whole new topic and should be treated separately.

Agreed! Should I create a new issue for that?

**Comment by bryan-brancotte** [comment]
TL;DR: show a message when it fails, don't check the schema for now : too much work.

The file is somehow checked, as we at least try to parse it. Indeed in case of error the app is in an obscure state and do not inform the user which is clearly bad. Checking is the json format is the one expected would be even better, but I think the first step is to inform the user that something went wrong. We can continu in this issue (I don't want to "loose" example and discussion), I renamed it to focus on what remains to do : have clear error message and not awkward state.

Validating the schema is the next step, we can for that do another issue as it requires more work, and deserve a discussion on whether it would be useful or not. In my case I don't think so as we do not have users using custom ontology (yet?).

**Comment by secrashi** [comment]
Ohhk, thank you so much for letting me know the in n outs of this. I'll soon make a PR for displaying the error!
